### ISSUE TITLE
[extension-upload] - unable to upload a brackets extension to the registry

### ISSUE THREAD

**Issue by timoweiss** [adobe/brackets#7387]
Hey, tried to upload an extension to the registry but i always get the following error:
"The extension package is invalid: The uploaded content is not a valid zip file."

My zip file has a size of about 3.1Mb.
It has a main.js and a valid package.json.
I have the suspicion that the large amount of files cause the error. Is it possible that there is any limit?

I created the zip via terminal 'zip requireRelease.zip . -r'

See also: https://groups.google.com/forum/#!topic/brackets-dev/srV4w5O0r94

**Comment by ingorichter** [comment]
I'm not sure if this might cause this issue. We have a fixed set of routes and nothing get's added dynamically.
We are dealing with a larger json object. If somebody adds or updates an extension, we are going to modify this json object. This might be more the issue I guess. But I'm not 100% sure about this either. ;-)

**Comment by MarcelGerber** [comment]
I don't think it's an issue with the JSON file.
The error message is "The uploaded content is not a valid zip file.", which is only ever reported if DecompressZip fails. So it looks like an update may help.

**Comment by peterflynn** [comment]
@MarcelGerber But with the same exact zip file, it succeeds after rebooting the server.  So there's a strong chance some sort of leak is at play here (memory leak, file handle leak, etc.).  That _could_ be due to a bug in the zip library, but it could also be a bug in any other part of our server, which leaks badly enough to eventually prevent DecompressZip from working properly.

That said, if there's a new version with bug fixes it seems worth upgrading just in case that helps...

**Comment by albertinad** [comment]
Hi @ingorichter , I'm still having problems since yesterday when trying to upload my package to the registry. I tried different zip packages, with some compression, without compression, the one that github generates (and then add the node modules I need), and the result is the same "The extension package is invalid: The uploaded content is not a valid zip file".  As a response of the /upload API I got a 404. The packages I tried are about 7,5mb and 9mb. I noticed that between yesterday and today that some packages were uploaded, I don't know if the issue can be related to bigger packages. Here is the link to one of the packages if you need it:  https://drive.google.com/file/d/0B2qjL2i00DnjUTR0YTVwblBtUGM/view?usp=sharing

**Comment by albertinad** [comment]
Hi all, I'm still having problems when uploading brackets-bower extension. After the last time, randomly it works, and now with the new version, that only is about 4.3mb,  I removed more files under node_modules/ inside bower module, and running npm depune, I'm again having the same issue "The extension package is invalid: The uploaded content is not a valid zip file".

**Comment by zaggino** [comment]
To add to the discussion:

I had problems uploading extensions packed together with https://github.com/gruntjs/grunt-contrib-compress so I switched to https://github.com/twolfson/grunt-zip and it all works fine since then.

`grunt-zip` uses different compression method than `grunt-contrib-compress` and the resulting zip file is much bigger with `grunt-zip` when compared to `grunt-contrib-compress`

From my case, I suspect that the registry is not able to work with some types of zip files, even though software that I use (winrar, 7zip) had no issues with any of them.

**Comment by ingorichter** [comment]
@zaggino yes, that is something that I have encountered too. I have mentioned this at a certain point, but some people didn't take it seriously. I got the answer, that this can not the reason for the upload failure. I have changed my projects to use grunt-zip too and since then I didn't have any issues anymore.
The remaining question for me is, if the generated zip file is valid. I started to investigate a bit in this direction and why we fail. It seems that the library that we use is expecting a specific flag in the zip file structure, but when generated with grunt-compress is not there. I have updated the zip library on my local machine to the latest version, but the error remains. I think the zip file is not valid. But I need some more proof to file appropriate bugs.

**Comment by peterflynn** [comment]
@ingorichter What library does grunt-contrib-compress itself use?  If that same library has an unzip API too, maybe we should look into switching to that lib.

It's beginning to seem like every Node zip library has bugs with certain files though -- iirc we've already switched libraries once because the original one didn't handle some files right.  This might be a whack-a-mole game where switching to grunt-compress's library fixes the zip files we're discussing here but breaks with other zip files that currently work fine :-(

**Comment by peterflynn** [comment]
Fwiw, looks like grunt-compress uses [archiver](https://github.com/archiverjs/node-archiver), which in turn wraps [zip-stream](https://github.com/archiverjs/node-zip-stream) -- but neither offers an unzip API...

**Comment by MarcelGerber** [comment]
I just looked for often-used zip libraries in npm and found [decompress-zip](https://www.npmjs.com/package/decompress-zip) (the one we currently use) and [adm-zip](https://www.npmjs.com/package/adm-zip).
We may want to try the latter.

@peterflynn Yeah, we used to use [unzip](https://www.npmjs.com/package/unzip) until https://github.com/adobe/brackets/commit/b69040a30d4979b8e0e1b9af19f25f6cfb78bb7f.

**Comment by albertinad** [comment]
To add to the discussion, my extension, [brackets-bower](https://github.com/albertinad/brackets-bower) uses the grunt tasks https://github.com/twolfson/grunt-zip but now with deflate compression, since I need to compress due to a lot of nested npm modules that bower has. Some time ago, it works fine, I was able to upload my extension to the brackets registry. But since a few months ago I'm having this same issue related to the registry, I've been trying to upload a zip file generated with the grunt-zip task, with and without deflate compression until my package was bigger than 10mb, then using windows winzip and 7zip, then using zip from linux, then git archive and downloading the zip file from github and manually adding the node modules I need. Then randomly I can upload my package, I don't know if that matches the moment when the server was restarted. I recently use an npm module, (flatten-packages)[https://github.com/arifsetiawan/flatten] to flatten the node modules tree  and then generate my zip package using grunt-zip and then zip from the command line and then winzip from windows, but the result was the same: the package is not a valid zip. So, I've updated the package.json and remove the node_modules folder, generate the zip file using grunt-zip task, and upload that package to the registry successfully as a test, then I deleted it. 
I think that the problem is related with packages that contains a lot of files or nested folders...

**Comment by zaggino** [comment]
@albertinad you're right about the nested folders/files, I've also seen this when trying to upload an extension which contains node_modules. One solution here would be https://github.com/adobe/brackets/pull/10602 , the other one I can think of is that we could start using npm as a registry, similarly to the jquery which also migrated from it's own registry to npm a while ago (http://plugins.jquery.com/)

**Comment by dnbard** [comment]
I still can reproduce this without any Grunt plugins, but with archives from "Download as ZIP" feature at GitHub. This issue are reproducible from time to time (not always).

**Comment by DenisVuyka** [comment]
I've started receiving 'Bad Request' today. Previously uploads worked fine for me.

**Comment by peterflynn** [comment]
@DenisVuyka That was a different problem that should be resolved now; see #10720.

**Comment by caphodel** [comment]
I've receiving 'Bad Request' today. Anything i can do to my zip file? or just wait someone fix this on the server side?

**Comment by phoenix3008** [comment]
I am trying to update my extension to a new version, and also get 400 'Bad request'

**Comment by mitchellsimoens** [comment]
Am also getting Bad Request. @peterflynn any way to kick the bucket again to fix?

**Comment by caphodel** [comment]
Now i got ENOSPC, that's mean there is no space left on the server.

**Comment by phoenix3008** [comment]
when will be some free space on server, to upload extensions ?

**Comment by nethip** [comment]
We are looking into. We have already sent a mail to aboutweb guys and hopefully this should get resolved sooner.

**Comment by caphodel** [comment]
Thank you so much @nethip

**Comment by maityarkodeb** [comment]
Subscribed to thread for future updates on this issue.

**Comment by zaggino** [comment]
Hi guys, some people in this thread might be interested:
https://groups.google.com/forum/#!topic/brackets-dev/lxMADzwDoWk

**Comment by dag0310** [comment]
Uploading my [Tab Out](https://github.com/dag0310/tab-out) extension to the [Extension Registry](https://brackets-registry.aboutweb.com/) was not possible yesterday ("Bad Request"), but today it worked. I didn't upload the same file, though. Today I used the GitHub release ZIP and yesterday I just tried uploading the GH master ZIP and using the git archive command.

**Comment by MiguelCastillo** [comment]
@dag0310 There were issues with the registry service, but they have been addressed.  So that's probably why it worked for you rather than the different artifact you were uplaoding.

@zaggino I have definitely thought I about doing the same thing. I will check out what you have...

**Comment by dnbard** [comment]
Same issue - #10832 reported again:

```
An error occurred, but unfortunately I don't know how to say this nicely: 
ENOSPC, mkdir '/home/aboutweb/tmp/bracketsPackage_11536-1774-qsodmm'
```

**Comment by smiclea** [comment]
I get "The extension package is invalid: The uploaded content is not a valid zip file." when I try to upload my extension. The upload works if I delete node_modules folder, but my extension requires it.

**Comment by ingorichter** [comment]
@smiclea how did you create the zip? Can I download it somewhere? Thanks.

**Comment by albertinad** [comment]
Hi @ingorichter , after a lot of failure intents uploading brackets-bower extension to the registry, I was able to reproduce the same error in my local environment. Below the steps I follow:
1) set up my local brackets registry
2) upload my package
    2.1) everything was right
3) clean up to start again
4) download 70 random extensions from the public brackets registry
5) start the local brackets registry
6) drag and drop all of the 70 extensions, wait until all of them where successfully uploaded
7) upload my package
    7.1) it fails, with the code that indicates that is not a vaild zip (which it is)

After digging a little and doing some debugging, I found the error code of it: EMFILE from the decompress-zip module/node. Related to bower/decompress-zip#26

Then, I updated the decompress-zip module to latest stable version, from `0.0.2` to `0.1.0`, repeat all the steps above multiple times, and I was able to upload to my local brackets registry my extension.
Also updated it on brackets and the tests passed.

We are using an outdated version of decompress-zip module, maybe it is time to update it.

**Comment by zaggino** [comment]
Good job @albertinad :+1:

**Comment by ingorichter** [comment]
Great job @albertinad. I suspected that decompress-zip might be the issue. I did some investigation some time ago but never finished the work (https://github.com/adobe/brackets/issues/7387). Sometimes the error disappeared after restarting the registry (I have not really an idea why).

**Comment by smiclea** [comment]
@ingorichter You can download it here: https://drive.google.com/file/d/0BxNY66eFHRb5YXI5UmlteVo2N0U/view?usp=sharing

**Comment by smiclea** [comment]
So it still doesn't work for me

**Comment by smiclea** [comment]
It works now. Thank you!

**Comment by kooohei** [comment]
I have same problem from today...

**Comment by kooohei** [comment]
this is my stuff.
http://dscs.jp/brackets-synapse.zip

**Comment by kooohei** [comment]
sorry I can't want, I have been required npm 's lodash and moment.
it works when cancled that.

thx.

**Comment by acbarbosa1964** [comment]
Dear friends,

I have several extensions I can upload, including previous version of the one I'm facing this same problem. I don't now what happened, I just change the project version and a string in my project and now I can't upload :(

It falls wit message:

acb-tartan-background-master.zip
1.5 MB
✘
The extension package is invalid: The uploaded content is not a valid zip file.

**Comment by swmitra** [comment]
@acbarbosa1964 if you have the source code for your extension uploaded to github, use 'download zip' utility from github project page and upload the same to registry. 

One more thing to validate is the package.json structure. Make sure it's not broken after the last update and still a valid json.

**Comment by acbarbosa1964** [comment]
Dear Mitra,

Thank you, but I have done that several time and I'm very sure JSON is
correct once I have used same structure for all my extensions.

Attached you have the zip file if you want to take a look.

Best Regards,


Thank you.




Atenciosamente,


*Antonio Carlos Barbosa*
*Fone:    *+55 15 2102-6500
---------------------------------------------
*http://www.alphatour.com.br <http://www.alphatour.com.br>*
*http://www.alphatour.info <http://www.alphatour.info>*

Um pensamento na madrugada...



*"Muitas vezes, as melhores coisas que temos a dizer para quem amamos,
temos como amigo ou nos relacionamos profissionalmente, são tão boas que
não conseguimos coloca-las para fora, mas as coisas ruins como críticas,
descontentamentos, raiva ..., que não queremos dentro de nós, são mais
fáceis de externalizar." Ainda hoje aprendi que dividindo estes sentimentos
bons que ficam guardados dentro do nosso peito, estamos na realidade
criando um efeito multiplicador, pois eles voltam para nós em dobro.*

Antonio Carlos Barbosa


*​*
A thought in the early hours ...

*"Often, the best things we have to say to those we love, we have as friend
or relate professionally, are so good that we couldn't put them out, but
the bad things like criticism, discontent, anger ... that we don't want
within us, are easier to speak out. *
*"*
Today I've learned that sharing these good feelings we have stored within
our chest, we are actually creating a multiplier effect, because they come
back to us twice.


- Antonio Carlos Barbosa


2015-10-19 15:22 GMT-02:00 Swagatam Mitra <notifications@github.com>:

**Comment by petetnt** [comment]
I am running into this issue with my new extension https://github.com/petetnt/brackets-sass-lint

I am getting `
{"errors":["The extension package is invalid:","The uploaded content is not a valid zip file."]}` despite being able to install it from URL or `.zip` with no issues. My package is `4.7 megabytes`

**Comment by swmitra** [comment]
@petetnt I have faced the same problem while updating my extension last year. Although I am not sure about the actual cause but the following steps got rid of the problem. 
1) Logout from extension registry 
2) Clear browser cache ( Not sure if mandatory but I did )
3) Close your browser 
4) Open registry again and then log in 
5) Try to upload the extension ( It worked for me !)

**Comment by mitchellsimoens** [comment]
I've also started using the [brackets-npm-registry](https://github.com/zaggino/brackets-npm-registry). It did take a full day for it to show up but publish to npm with `brackets-extension` in the `keywords` array of `package.json` and it picked it up. To update, just issue a `npm publish` and the registry picks up the update.

Would be great to migrate to this instead of the builtin one that I've had issues with it going down (has been a while for that tho).

**Comment by petetnt** [comment]
@swmitra Good idea, but sadly those steps wouldn't work for me. :crying_cat_face:

**Comment by petetnt** [comment]
Confirmed that it was the `node_modules` that are the culprit: there's a lot of them (3387 Files, 547 Folders) but the combined size is still less than ~4 megabytes. Removing them from the zip file made the process a success. 

I tried to pack the modules with `npm@v2` but that didn't work either: only thing it did was was get the package size up to 7 megabytes.

The thing is that I can't even lower my amount of files: I only have one external dependency, `sass-lint` that just happens to use tons of external dependencies :crying_cat_face:

**Comment by Intronet** [comment]
The extensions registry upload is not working again, and I see it is nearly a full year that this topic was opened -  I need to upload v0.0.5
Can you restart the server please
[[Brackets Website Admin]](https://s3.amazonaws.com/extend.brackets/brackets-website-admin/brackets-website-admin-0.0.4.zip)

**Comment by alemonteiro** [comment]
I'm unable to upload an new version also. None of the work around  suggestions worked for me.
The new package: http://alemonteiro.com.br/brackets-sql-connector.zip

**Comment by alemonteiro** [comment]
Still can't upload the update for sql-connector but was able to upload a new smaller extension.
The connector one tried from linux using file roller, zip and rar,  and from windows using winrar e 7zip, none worked.
![screenshot from 2016-04-12 19 40 36](https://cloud.githubusercontent.com/assets/2397555/14478061/a6828cd8-00e8-11e6-8389-c21c54bbaca9.png)

**Comment by alemonteiro** [comment]
Maybe this issue is related to file size? I couldn't upload anything larger than 2mb but never had problem with small packages.

So added auto install of node dependencies on the extension so the package would be small and only then was able to upload and update the version on registry.

**Comment by dnbard** [comment]
@alemonteiro I cannot upload 99kb extension right now so it's not related to size, I think.

**Comment by ingorichter** [comment]
I always had the impression that it was related to the program that created the zip file. I don't have 100% proof for it, but it seems that the zip library used to decompress the uploaded zip chokes on some of the files.

**Comment by alemonteiro** [comment]
@dnbard , i don't think it's only size related, but it may play a role...

@ingorichter , i had this impression also, but it got mixed up with the size thing...

Using wirar over windows I never had a problem with upload, did various uploads until 1.5mb.
But some weeks ago I migrated to Debian and the first package that I tried to upload, zipped with archiver, wasn't accepted. So I tried using xarchiver, rar and zip directly, but with no luck.
Then I got this same package, about 2mb, and tried from windows (with Winrar and 7zip) but also didn't get accepted.

I could only get it to be accepted again after removing the dependecies from the package, so it would be like 200kb, and with that size any zipper that i tried worked.

I think this won't help much because it feels inconsistent. I'll try to run a version locally to debug it once I get some free time.
